internal/pkg/skuba/replica/replica.go

```diff
@@ -0,0 +1,264 @@
+		ClusterSize:       len(node.Items),
+		MinSize:           minSize,
+		SelectDeployments: *fromSelectors,
+	}, nil
+}
+
+func (r *Helper) replaceAffinity(remove string, create string) error {
+	// doRemove will remove only when affinity path exists
+	doRemove := false
+	affinity := r.Deployment.Spec.Template.Spec.Affinity
+	if affinity != nil && affinity.PodAntiAffinity != nil {
+		preferredList := len(r.Deployment.Spec.Template.Spec.Affinity.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution)
+		requiredList := len(r.Deployment.Spec.Template.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution)
+		switch {
+		case remove == "preferred":
+			if preferredList != 0 {
+				doRemove = true
+				break
+			}
+			if requiredList != 0 {
+				return nil
+			}
+		case remove == "required":
+			if requiredList != 0 {
+				doRemove = true
+				break
+			}
+			if preferredList != 0 {
+				return nil
+			}
+		}
+	}
+	if doRemove {
+		if err := r.removeAffinity(); err != nil {
+			return err
+		}
+	}
+
+	affinityJSON := fmt.Sprintf(patchAffinityPreferred, r.Deployment.ObjectMeta.Name)
+	if create == "required" {
+		affinityJSON = fmt.Sprintf(patchAffinityRequired, r.Deployment.ObjectMeta.Name)
+	}
+	_, err := r.Client.AppsV1().Deployments(metav1.NamespaceSystem).Patch(r.Deployment.ObjectMeta.Name, types.StrategicMergePatchType, []byte(affinityJSON))
+	if err != nil {
+		return err
+	}
+
+	r.isUpdated = true
+	return nil
+}
+
+func (r *Helper) removeAffinity() error {
+	_, err := r.Client.AppsV1().Deployments(metav1.NamespaceSystem).Patch(r.Deployment.ObjectMeta.Name, types.JSONPatchType, []byte(patchAffinityRemove))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Helper) updateDeploymentReplica(size int) (*appsv1.Deployment, error) {
+	replicaJSON := fmt.Sprintf(patchReplicas, size)
+	return r.Client.AppsV1().Deployments(metav1.NamespaceSystem).Patch(r.Deployment.ObjectMeta.Name, types.StrategicMergePatchType, []byte(replicaJSON))
+}
+
+func (r *Helper) refreshReplicas() error {
+	// at least one node needs to be free of the affinity rules for pods to re-distribute.
+	// sizing down replicas would make room for new pod resource.
+	if _, err := r.updateDeploymentReplica(r.ClusterSize - 1); err != nil {
+		return err
+	}
+	if _, err := r.updateDeploymentReplica(r.ReplicaSize); err != nil {
+		return err
+	}
+	return nil
+}
+
+// UpdateNodes patches for replicas affinity rules after adding of nodes.
+// This updates affinity rule to preferredDuringSchedulingIgnoredDuringExecution of
+// cluster size less then replica size,
+// And updates affinity rule to requiredDuringSchedulingIgnoredDuringExecution when
+// cluster size is equal or greater to replicas.
+func (r *Helper) UpdateNodes() error {
+	for _, deployment := range r.SelectDeployments.Items {
+		r.Deployment = deployment
+		r.ReplicaSize = int(*r.Deployment.Spec.Replicas)
+		if r.ReplicaSize == 0 {
+			r.ReplicaSize = r.MinSize
```

Comment:
It would be good add unit test to cover this case.

Reply:
Thanks, missed this. Covered